src/Microsoft.AspNetCore.Mvc.TagHelpers/Cache/HybridHtmlFragmentCache.cs

```diff
@@ -0,0 +1,217 @@
+        private const string DistributedOptionName = "distributed";
+
+        private readonly IMemoryCache _memoryCache;
+        private readonly IDistributedCache _distributedCache;
+
+        public HybridHtmlFragmentCache(IMemoryCache memoryCache)
+        {
+            _memoryCache = memoryCache;
+        }
+
+        public HybridHtmlFragmentCache(IMemoryCache memoryCache, IDistributedCache distributedCache)
+        {
+            _memoryCache = memoryCache;
+            _distributedCache = distributedCache;
+        }
+
+        public Task<IHtmlContent> SetAsync(string key, Func<Task<IHtmlContent>> renderContent, HtmlFragmentCacheContext context)
+        {
+            // is the distributed option set?
+            if (EnsureDistributedService(context))
+            {
+                return SetDistributedAsync(key, renderContent, context, _distributedCache);
+            }
+            else
+            {
+                return SetMemoryAsync(key, renderContent, context);
+            }
+        }
+
+        private async Task<IHtmlContent> SetMemoryAsync(string key, Func<Task<IHtmlContent>> renderContent, HtmlFragmentCacheContext context)
+        {
+            var options = new MemoryCacheEntryOptions();
+
+            if (context.ExpiresOn != null)
+            {
+                options.SetAbsoluteExpiration(context.ExpiresOn.Value);
+            }
+
+            if (context.ExpiresAfter != null)
+            {
+                options.SetAbsoluteExpiration(context.ExpiresAfter.Value);
+            }
+
+            if (context.ExpiresSliding != null)
+            {
+                options.SetSlidingExpiration(context.ExpiresSliding.Value);
+            }
+
+            if (context.HasOptions && context.Options.ContainsKey(PriorityOptionName))
+            {
+                CacheItemPriority priority;
+                if (Enum.TryParse(context.Options[PriorityOptionName], out priority))
+                {
+                    options.SetPriority(priority);
+                }
+            }
+
+            // Create an entry link scope and flow it so that any tokens related to the cache entries
+            // created within this scope get copied to this scope.
+            using (var link = _memoryCache.CreateLinkingScope())
+            {
+                var content = await renderContent();
+
+                options.AddEntryLink(link);
+
+                var stringBuilder = new StringBuilder();
+                using (var writer = new StringWriter(stringBuilder))
+                {
+                    content.WriteTo(writer, context.HtmlEncoder);
+                }
+
+                var result = new StringBuilderHtmlContent(stringBuilder);
+
+                return _memoryCache.Set<IHtmlContent>(key, result, options);
+            }
+        }
+
+        public async Task<IHtmlContent> SetDistributedAsync(string key, Func<Task<IHtmlContent>> renderContent, HtmlFragmentCacheContext context, IDistributedCache cache)
```

Comment:
Should do the same elsewhere, looks like there's too many `public` methods in this class